jupyterhub/assignments/intake-dashboard/report_districts.ipynb

```diff
@@ -0,0 +1,139 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%capture\n",
+    "%load_ext sql\n",
+    "%config SqlMagic.feedback = False\n",
+    "%config SqlMagic.displaylimit = 0\n",
+    "%config SqlMagic.displaycon = False"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import datetime\n",
+    "\n",
+    "start = datetime.datetime.now() - datetime.timedelta(days=90)\n",
+    "end = datetime.datetime.now()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%sql report_districts <<\n",
+    "SELECT \n",
+    "    r.district as district,\n",
+    "    r.primary_statute as classification,\n",
+    "    r.primary_complaint as complaint,\n",
+    "    r.location_state as state,\n",
+    "    COUNT(*) count\n",
+    "FROM cts_forms_report r\n",
+    "WHERE district notnull AND primary_complaint notnull\n",
+    "GROUP BY complaint, district, classification, state\n",
+    "HAVING COUNT(*) > 1\n",
+    "ORDER BY count desc"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from IPython.display import display, HTML\n",
+    "import numpy as np\n",
+    "primary_complaint_choices = {\n",
+    "    'workplace': 'Workplace discrimination or other employment-related problem',\n",
+    "    'housing': 'Housing discrimination or harassment',\n",
+    "    'education': 'Discrimination at a school, educational program or service, or related to receiving education',\n",
+    "    'police': 'Mistreated by police, correctional staff, or inmates',\n",
+    "    'voting': 'Voting rights or ability to vote affected',\n",
+    "    'commercial_or_public': 'Discriminated against in a commercial location or public place',\n",
+    "    'something_else': 'Something else happened',\n",
+    "}\n",
+    "readable_start = start.strftime(\"%B %d, %Y\")\n",
+    "readable_end = end.strftime(\"%B %d, %Y\")\n",
+    "report_district_data = report_districts.DataFrame()\n",
+    "districts = report_district_data['district'].values\n",
+    "data_divs = []\n",
+    "district_dropdown = []\n",
+    "for i, district in enumerate(districts):\n",
```

Comment:
A couple of (optional) tips in the vein of functional, immutable code:
- Refactor the count controlled loop using `zip`
- Use a helper function to build the data_divs list with a comprehension instead of `append`

you might be able to use `zip` instead of `enumerate` to avoid the need for a counter here.

```py
complaints = [
  primary_complaint_choices.get(complaint_choice)
  for complaint_choice
  in report_district_data['complaint'].values
]
classifications = report_district_data['classification'].values
states = report_district_data['state'].values
counts = report_district_data['count'].values
records = zip(districts, complaints, classifications, states, counts)
data_divs = [
  make_data_div(district, complaint, classification, state, count)
  for district, complaint, classification, state, count in records
]

def make_data_div(district, complaint, classification, state, count):
  return f"""that multiline string"""


```

Reply:
(this pattern can also apply to the other file)